java/code/src/com/redhat/rhn/manager/contentmgmt/ContentManager.java

```diff
@@ -856,18 +850,44 @@ private static void alignErrata(Channel src, Channel tgt, Collection<ErrataFilte
         ErrataManager.mergeErrataToChannel(user, includedErrata, tgt, src, false, false);
     }
 
-    private static List<Long> extractPackageIds(Collection<Package> packages) {
-        return packages.stream().map(p -> p.getId()).collect(toList());
-    }
+    /**
+     * Filters given entities based on given filters.
+     *
+     * Returns a Pair containing:
+     * - Left side: set of entities that are kept (not filtered out)
+     * - Right side: set of entities that are filtered out
+     *
+     * Entities are processed one-by-one by filters as follows:
+     * - when any DENY filter is satisfied for an entity -> this entity gets filtered out
+     * - when an ALLOW is satisfied for an entity -> this entity gets NOT filtered out (even if it had been filtered out
+     *   by a DENY filter = ALLOW filters have higher priority)
+     *
+     * @param entities entities, e.g. packages
+     * @param filters filters, e.g. package filters
+     * @param <T> the type of the entity (e.g. Package)
+     * @return Pair containing (left side) a set of entities not filtered-out
+     * and (right side) a set of entities filtered out
+     */
+    private static <T> Pair<Set<T>, Set<T>> filterEntities(Set<T> entities,
+            Collection<? extends ContentFilter<T>> filters) {
+        Map<ContentFilter.Rule, List<ContentFilter<T>>> filtersByRule = filters.stream()
+                .collect(groupingBy(ContentFilter::getRule));
 
-    private static Set<Package> filterPackages(Set<Package> packages, Collection<PackageFilter> packageFilters) {
-        Predicate<Package> compositePredicate = packageFilters.stream()
+        Predicate<T> denyPredicate = filtersByRule.getOrDefault(ContentFilter.Rule.DENY, emptyList()).stream()
                 .map(f -> (Predicate) f)
-                .reduce((f1, f2) -> f1.and(f2))
-                .orElse(p -> true);
-        return packages.stream()
-                .filter(compositePredicate)
-                .collect(toSet());
+                .reduce(x -> false, (f1, f2) -> f1.or(f2));
+
+        Predicate<T> allowPredicate = filtersByRule.getOrDefault(ContentFilter.Rule.ALLOW, emptyList()).stream()
+                .map(f -> (Predicate) f)
+                .reduce(x -> false, (f1, f2) -> f1.or(f2));
+
+        Predicate<T> compositePredicate = denyPredicate.negate().or(allowPredicate);
```

Comment:
As per the `allow` condition *overrides* the `deny` one because it has more power, wouldn't it make sense to apply the `allow` criteria first in order to have the `condition check shortcut if true` instead of computing the `deny` on every item?

Reply:
From the performance point of view, it'd better to arrange the predicates in the composite predicate in that way, that the first one to evaluate is more likely to return `true`, thus avoiding unnecessary evaluation of the second one. The problem with this is that we don't know, which one of them is that :)

Consider a content project with 2 attached filters:
1. filter `ALLOW`ing errata `SUSE-CL-123`
2. filter `DENY`ing errata issued after `2019-01-01`

The current implementation goes through all errata and for each of them it checks:
- is `denyPredicate.negate()` true? if so -> include the erratum in the result. This happens for a lot of errata (the ones issued before `2019-01-01`) and for these the next step is not necessary.
- if the previous point is false -> check the `allowPredicate` is true (check if the erratum is `SUSE-CL-123`).

In this scenario, we avoid some `allowPredicate` evaluations.

If we flipped the evaluation of the allow/deny filters, we'd end up for evaluating the `allowPredicate` for almost all the errata (except `SUSE-CL-123`) and we'd also need to evaluate the `denyPredicate` for allmost all errata (since they didn't pass the `allowPredicate`). This wouldn't be that good.

On the other hand, if you have a scenario when the `ALLOW` filters return true for a lot of errata, it'd be better to evaluate the `allowPredicate` first, but how do you know? You can craft multiple scenarios where one way is better then the other and vice versa, so I think the bottom line is that it doesn't really matter, in which order you evaluate them.